Sources/NIO/ChannelPipeline.swift

```diff
@@ -944,7 +957,163 @@ extension ChannelPipeline {
+        ///
+        /// - Parameters:
+        ///   - handlers: The handlers to add.
+        ///   - position: The position in the `ChannelPipeline` to add `handlers`. Defaults to `.last`.
+        public func addHandlers(_ handlers: ChannelHandler...,
+                                position: ChannelPipeline.Position = .last) throws {
+            try self.pipeline.addHandlersSync(handlers, position: position).get()
+        }
+
+        /// Returns the `ChannelHandlerContext` for the given handler instance if it is in
+        /// the `ChannelPipeline`, if it exists.
+        ///
+        /// - Parameter handler: The handler belonging to the context to fetch.
+        /// - Returns: The `ChannelHandlerContext` associated with the handler.
+        public func context(handler: ChannelHandler) throws -> ChannelHandlerContext {
+            return try self.pipeline._contextSync({ $0.handler === handler }).get()
+        }
+
+        /// Returns the `ChannelHandlerContext` for the handler with the given name, if one exists.
+        ///
+        /// - Parameter name: The name of the handler whose context is being fetched.
+        /// - Returns: The `ChannelHandlerContext` associated with the handler.
+        public func context(name: String) throws -> ChannelHandlerContext {
+            return try self.pipeline.contextSync(name: name).get()
+        }
+
+        /// Returns the `ChannelHandlerContext` for the handler of given type, if one exists.
+        ///
+        /// - Parameter name: The name of the handler whose context is being fetched.
+        /// - Returns: The `ChannelHandlerContext` associated with the handler.
+        public func context<Handler: ChannelHandler>(handlerType: Handler.Type) throws -> ChannelHandlerContext {
+            return try self.pipeline._contextSync(handlerType: handlerType).get()
+        }
+
+        /// Returns the `ChannelHandler` of the given type from the `ChannelPipeline`, if it exists.
+        /// - Returns: A `ChannelHandler` of the given type if one exists in the `ChannelPipeline`.
+        public func handler<Handler: ChannelHandler>(type _: Handler.Type) throws -> Handler {
+            return try self.pipeline._handlerSync(type: Handler.self).get()
+        }
+    }
+
+    /// Invokes a closure with a synchronous view of this `ChannelPipeline` if this function is
+    /// called from the `EventLoop`.
+    ///
+    /// - Important: Users *must not* escape the `SynchronousView` from the closure.
+    /// - Parameter closure: A closure to operate on a synchronous view of this `ChannelPipeline`.
+    /// - Throws: `NIONotOnEventLoopError` if this function is not invoked on the event loop for
+    ///     this `ChannelPipeline`. Re-throws any error thrown from `body`.
+    public func withSynchronousViewIfAvailable(_ closure: (SynchronousView) throws -> Void) throws {
+        if self.eventLoop.inEventLoop {
+            try closure(SynchronousView(_pipeline: self))
```

Comment:
So technically, this is incorrect. `self.eventLoop.inEventLoop` must only be used as an _optimisation_, not as a decider. In NIOTS, this may say `false` even if you're on it.